src/transformers/quantizers/quantizer_hqq.py

```
@@ -0,0 +1,199 @@
+
+
+if is_accelerate_available():
+    from accelerate.hooks import remove_hook_from_module
+
+if is_torch_available():
+    import torch
+
+logger = logging.get_logger(__name__)
+
+
+class HQQHfQuantizer(HfQuantizer):
+    """
+    HQQ quantizer base HF class.
+    nn.Linear modules are first tagged with quant_config in _process_model_before_weight_loading().
+    The actually quantization and offloading to the GPU is done in check_quantized_param().
+    self.show_progress (bool) is used to show quantization progress in each shard.
+    """
+
+    use_keep_in_fp32_modules = False
+    requires_parameters_quantization = True
+    requires_calibration = False
+    required_packages = ["hqq"]
+
+    def __init__(self, quantization_config, **kwargs):
+        super().__init__(quantization_config, **kwargs)
+        self.show_progress = quantization_config.show_progress
+        self.torch_dtype = None
+        self.using_multi_gpu = False
+
+    def validate_environment(self, *args, **kwargs):
+        if not (is_hqq_available()):
+            raise ImportError(
+                "HQQ is not available. Please follow the instructions to install it: `https://github.com/mobiusml/hqq/`"
+            )
+
+        if kwargs.get("from_tf", False) or kwargs.get("from_flax", False):
+            raise ValueError(
+                "Converting weights from tf/flax weights is currently not supported, please make"
+                " sure the weights are in PyTorch format."
+            )
+
+        if not torch.cuda.is_available():
+            raise RuntimeError("No GPU found. A GPU is needed for quantization.")
+
+        if self.torch_dtype is None:
+            self.torch_dtype = kwargs.get("torch_dtype", torch.float16)
+
+        if self.using_multi_gpu is False:
+            if "device_map" in kwargs:
+                if isinstance(kwargs["device_map"], dict):
+                    self.using_multi_gpu = len({item[1] for item in kwargs["device_map"].items()}) > 1
```

Before that, we should probably check also that the `device_map` doesn't not contain `cpu` or `disk` since need to initialize everything on the GPUs.

---

Something like this? https://github.com/huggingface/transformers/pull/29637/commits/20f9ad5bfa584064c1ab00ec60f288aa64324e76

---

I mean does HQQ work with cpu/disk offload ? if not, we need to return an error saying that we can't load the model.

---

But there's already this check (following bnb):
https://github.com/huggingface/transformers/pull/29637/files#diff-47d0e774fd1df1be9bfb94cf01ada73487310534be6c88191c92ea18e1d274d2R77-R78

---

Yes but that's not enough. With accelerate, some layers of the model can also be offloaded to the cpu. See how to check for that [here](https://github.com/huggingface/transformers/blob/20081c743ee2ce31d178f2182c7466c3313adcd2/src/transformers/quantizers/quantizer_eetq.py#L74).

---

I think this should do it, please confirm: https://github.com/huggingface/transformers/pull/29637/commits/5283ac2043c76d917247a845295bae0b30846a3a